tools/download_dart_sdk.py

```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env python3
+# Copyright 2013 The Flutter Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# This python script downloads the Dart SDK that matches the version in the
+# source tree and puts it in prebuilts/. The return code of this script will
+# always be 0, even if there is an error, unless the --fail-loudly flag is
+# passed.
+# TODO(zra): Eliminate this script and download through the DEPS file if/when
+# the Dart SDKs pulled by this script are uploaded to cipd.
+
+import argparse
+import os
+import multiprocessing
+import platform
+import re
+import shutil
+import subprocess
+import sys
+import zipfile
+
+
+SRC_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+FLUTTER_DIR = os.path.join(SRC_ROOT, 'flutter')
+FLUTTER_PREBUILTS_DIR = os.path.join(FLUTTER_DIR, 'prebuilts')
+DART_DIR = os.path.join(SRC_ROOT, 'third_party', 'dart')
+DART_VERSION = os.path.join(DART_DIR, 'tools', 'VERSION')
+
+
+# The Dart SDK script is the source of truth about the sematic version.
+sys.path.append(os.path.join(DART_DIR, 'tools'))
+import utils
+
+
+# Prints to stderr.
+def eprint(*args, **kwargs):
+  print(*args, file=sys.stderr, **kwargs)
+
+
+# Try to guess the host operating system.
+def GuessOS():
+  os_name = utils.GuessOS()
+  if os_name == 'win32':
+    os_name = 'windows'
+  if os_name not in ['linux', 'macos', 'windows']:
+    eprint('Could not determine the OS: "%s"' % os_name)
+    return None
+  return os_name
+
+
+# For os `os_name` return a list of architectures for which prebuilts are
+# supported. Kepp in sync with `can_use_prebuilt_dart` in //flutter/tools/gn.
+def ArchitecturesForOS(os_name):
+  if os_name == 'linux':
+    return ['x64', 'arm64']
+  elif os_name == 'macos':
+    return ['x64', 'arm64']
+  elif os_name =='windows':
+    return ['x64']
+
+  eprint('Could not determine architectures for os "%s"' % os_name)
+  return None
+
+
+# Downloads a Dart SDK to //flutter/prebuilts.
+def DownloadDartSDK(channel, version, os_name, arch):
+  file = 'dartsdk-{}-{}-release.zip'.format(os_name, arch)
+  url = 'https://storage.googleapis.com/dart-archive/channels/{}/raw/{}/sdk/{}'.format(
```

Comment:
This seems to be recreating the functionality of CIPD hooks in the DEPS file. Can we move these artifacts to CIPD instead? Stuff like needing curl on the host and not performing hash consistency checks may lead to unnecessary hassles on CI.

Reply:
Yes. Sorry for not being more explicit about the plan. I want to see how much this can improve build times on CI. If there's a reasonably good improvement, then I'll open a thread with Dart EngProd to request that these artifacts are uploaded to cipd, and this script will go away.